src/ucx_plugin.c

```diff
@@ -297,46 +301,60 @@ static ncclResult_t ucx_get_ctx_and_worker(int dev, ucp_context_h *ctx,
 
   ucp_worker_progress(*worker);
   workers[dev].count++;
-#else
-  ucx_init_context(ctx, dev);
-  ucx_init_worker(*ctx, worker);
-  if (newtag != NULL) {
-    *newtag = tag;
-  }
-#endif
   pthread_mutex_unlock(&nccl_ucx_lock);
   return ncclSuccess;
 }
 
 static ncclResult_t nccl_ucx_free_worker(ucp_worker_h worker) {
-  int i, dummy, count;
+  int i, dummy, count, done = 0;
   struct ep_list *ep, *cur;
+  struct nccl_ucx_worker *ucx_worker, *next;
+  ncclResult_t result;
 
   pthread_mutex_lock(&nccl_ucx_lock);
   for(i = 0; i < ncclNIbDevs; i++) {
     if (workers[i].count > 0 && worker == workers[i].worker) {
       count = --workers[i].count;
+      if (count == 0) {
+        workers[i].next = worker_gc_list;
+        worker_gc_list  = &workers[i];
+        worker_count--;
+        done = worker_count == 0;
+      }
       break;
     }
   }
   pthread_mutex_unlock(&nccl_ucx_lock);
 
-  if (i < ncclNIbDevs && count == 0) {
-    ep = workers[i].eps;
-    while(ep){
+  if (!done) {
+    return ncclSuccess;
+  }
+
+  for (ucx_worker = worker_gc_list; ucx_worker != NULL; ucx_worker = next) {
```

Comment:
do we really need `gc_list` or we can just iterate over device worker once free was invoked for all of them?

Reply:
agree, we could iterate over the `workers[]` but the plan is to remove that array altogether and allocate workers maybe rx/tx based + max number of ep per worker combination.

Reply:
ok